### ISSUE TITLE
Test failures in Fedora

### ISSUE THREAD

**Issue by pvalena** [Shopify/bootsnap#395]
### I made sure the issue is in bootsnap

The issue is with bootsnap test suite I've encountered while making a package upgrade for Fedora.

### Steps to reproduce

Tests are failing; executed with:
`ruby -rpathname -rmsgpack -rset -Ilib:test:ext -e 'Dir.glob "./test/**/*_test.rb", &method(:require)'`

To avoid dependency bloat during build, and for better transparency.

### Actual behavior
```
  1) Failure:
CompileCacheJSONTest#test_load_file_freeze
[/builddir/build/BUILD/bootsnap-1.10.1/usr/share/gems/gems/bootsnap-1.10.1/test/compile_cache/json_test.rb:65]:

Expected # encoding: UTF-8
#    valid: true
"foo" to be frozen?.
  2) Error:
Bootsnap::LoadPathCache::CacheTest#test_path_encoding:
TypeError: no implicit conversion of nil into String
    <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:85:in `require'
    <internal:/usr/share/rubygems/rubygems/core_ext/kernel_require.rb>:85:in `require'

/builddir/build/BUILD/bootsnap-1.10.1/usr/share/gems/gems/bootsnap-1.10.1/test/load_path_cache/cache_test.rb:179:in `test_path_encoding'

```

I also received error `NoMethodError: undefined method 'bundle_path' for Bundler:Module` in `test/cli_test.rb`, but I that is probably caused by the way we execute the test suite.

### System configuration

**Bootsnap version**:
1.10.1

**Ruby version**:
ruby 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [x86_64-linux]

**Rails version**:
7.0.0

**Comment by pvalena** [comment]
> I think the first one could be explained by an older message pack version. What's the version of your message pack gem?

msgpack (1.3.3)

> 
> The second one is very weird, I need to look more into it.

Thanks!

**Comment by casperisfine** [comment]
> msgpack (1.3.3)

Ok, so that explains it. `:freeze` is only supported on `msgpack 1.4.0`. It's not a huge deal but I'll fix the code to better handle this case.

**Comment by pvalena** [comment]
> @pvalena could you test this rev: [b249255](https://github.com/Shopify/bootsnap/commit/b249255b7953731586440fe680a541d83a505eeb) ?

```
  1) Failure:
Bootsnap::LoadPathCache::CacheTest#test_path_encoding [/builddir/build/BUILD/bootsnap-1.10.1/usr/share/gems/gems/bootsnap-1.10.1/test/load_path_cache/cache_test.rb:179]:
--- expected
+++ actual
@@ -1,3 +1 @@
-# encoding: UTF-8
-#    valid: true
-"/tmp/d20220117-1606878-yw6mbo/b\u00E9\u00E9.rb"
+nil
```

**Comment by casperisfine** [comment]
Ok, that's super weird. `"/tmp/d20220117-1606878-yw6mbo/b\u00E9\u00E9.rb"` suggest your Ruby has some encoding issues regarding UTF-8.

That string should be displayed like `d20220117-25758-rhl30x/béé.rb"`.

Is Fedora's Ruby tweaked in any way?

**Comment by casperisfine** [comment]
Ok, I can indeed repro the issue with:

```
$ env LC_ALL=C.UTF-8 bundle exec ruby -Itest test/load_path_cache/cache_test.rb -n test_path_encoding
Run options: -n test_path_encoding --seed 44785

# Running:

F

Finished in 0.023010s, 43.4594 runs/s, 217.2968 assertions/s.

  1) Failure:
Bootsnap::LoadPathCache::CacheTest#test_path_encoding [test/load_path_cache/cache_test.rb:179]:
--- expected
+++ actual
@@ -1,3 +1 @@
-# encoding: UTF-8
-#    valid: true
-"/private/var/folders/vy/srfpq1vn6hv5r6bzkvcw13y80000gn/T/d20220117-26363-19eg9ur/b\u00E9\u00E9.rb"
+nil
```

That's interesting, I'll dig into this. Either way, this test failing just mean Bootsnap will use very marginally more memory, not that it will be broken.

**Comment by casperisfine** [comment]
```bash
$ ruby -e 'p [Encoding.default_internal, Encoding.default_external]'
[nil, #<Encoding:UTF-8>]
$ env LC_ALL=C.UTF-8 ruby -e 'p [Encoding.default_internal, Encoding.default_external]'
[nil, #<Encoding:US-ASCII>]
```

**Comment by casperisfine** [comment]
Reported upstream: https://bugs.ruby-lang.org/issues/18495

I'll see if I can produce the Ruby fix, then I'll look at handling `Encoding.default_external == Encoding::US_ASCII` better in Bootsnap.

**Comment by casperisfine** [comment]
@pvalena I dug a bit more into it, and while not uncommon, `LC_ALL=C.UTF-8` isn't a valid POSIX value, and that's why Ruby is confused. I'd suggest not to set that variable for Ruby.

I'll still improve support for it in Bootsnap, but you are likely to encounter other issues in other Ruby libraries.

**Comment by pvalena** [comment]
Thanks for your help!

**Comment by casperisfine** [comment]
I merged 8ddb820468169fab77fd4d02dd6674313c6e4364, which skips this test if `Encoding.default_external != Encoding::UTF_8`.

I looked into fully supporting this case, but it would add a significant overhead for only a small memory usage reduction, so not worth it considering that `Encoding.default_external != Encoding::UTF_8` is likely to be a system config issue.

Thanks for the reports though, very helpful.
